gcc/config/riscv/rvp.md

```diff
@@ -0,0 +1,7058 @@
+		(sign_extend:DI
+		  (vec_select:HI
+		    (match_dup 2)
+		      (parallel [(match_operand:SI 4 "imm_2_3_operand"   "C02, C03")]))))]
+	      UNSPEC_KMMW)
+	      (const_int 15)))))]
+  "TARGET_ZPN && TARGET_64BIT"
+  "@
+   kmmwb2\t%0, %1, %2
+   kmmwt2\t%0, %1, %2"
+[(set_attr "type" "simd")])
+
+(define_expand "kmmwb2"
+  [(match_operand:SI 0 "register_operand" "")
+   (match_operand:SI 1 "register_operand" "")
+   (match_operand:V2HI 2 "register_operand" "")]
+  "TARGET_ZPN && !TARGET_64BIT"
+{
+  emit_insn (gen_kmmw2_internal (operands[0], operands[1], operands[2], GEN_INT (0)));
+  DONE;
+}
+[(set_attr "type" "simd")])
+
+(define_expand "kmmwt2"
+  [(match_operand:SI 0 "register_operand" "")
+   (match_operand:SI 1 "register_operand" "")
+   (match_operand:V2HI 2 "register_operand" "")]
+  "TARGET_ZPN && !TARGET_64BIT"
+{
+  emit_insn (gen_kmmw2_internal (operands[0], operands[1], operands[2], GEN_INT (1)));
+  DONE;
+}
+[(set_attr "type" "simd")])
+
+(define_insn "kmmw2_internal"
+  [(set (match_operand:SI 0 "register_operand"                       "=   r,    r")
+	(truncate:SI
+	  (lshiftrt:DI
+	    (unspec:DI [(mult:DI
+	      (sign_extend:DI (match_operand:SI 1 "register_operand" "    r,    r"))
+	      (sign_extend:DI
+		(vec_select:HI
+		  (match_operand:V2HI 2 "register_operand"           "    r,    r")
+		  (parallel [(match_operand:SI 3 "imm_0_1_operand"   "  C00,  C01")]))))]
+	    UNSPEC_KMMW)
+	    (const_int 15))))]
+  "TARGET_ZPN && !TARGET_64BIT"
+  "@
+   kmmwb2\t%0, %1, %2
+   kmmwt2\t%0, %1, %2"
+  [(set_attr "type" "simd")])
+;; KSLLW
+(define_insn "ksll"
+  [(set (match_operand:SI 0 "register_operand"               "=   r, r")
+	(ss_ashift:SI (match_operand:SI 1 "register_operand" "    r, r")
+		      (match_operand:SI 2 "rimm5u_operand"   " Iu05, r")))]
+  "TARGET_ZPN"
+  "@
+   kslliw\t%0, %1, %2
+   ksllw\t%0, %1, %2"
+  [(set_attr "type"  "dsp")
+   (set_attr "mode"  "SI")])
+
+;; KSLL8
+(define_insn "kslli8<VQI:mode><X:mode>"
+  [(set (match_operand:VQI 0 "register_operand"                "=  r, r")
+	(ss_ashift:VQI (match_operand:VQI 1 "register_operand" "   r, r")
+		       (match_operand:X 2   "rimm3u_operand"   " u03, r")))]
+  "TARGET_ZPN"
+  "@
+   kslli8\t%0, %1, %2
+   ksll8\t%0, %1, %2"
+  [(set_attr "type"  "simd, simd")
+   (set_attr "mode"  "<VQI:MODE>, <VQI:MODE>")])
+
+;; KSLL16
+(define_insn "kslli16<mode>"
+  [(set (match_operand:VHI 0 "register_operand"                "=  r, r")
+	(ss_ashift:VHI (match_operand:VHI 1 "register_operand" "   r, r")
+		       (match_operand:SI 2  "rimm4u_operand"   " u04, r")))]
+  "TARGET_ZPN"
+  "@
+   kslli16\t%0, %1, %2
+   ksll16\t%0, %1, %2"
+  [(set_attr "type"  "simd, simd")
+   (set_attr "mode"  "<MODE>, <MODE>")])
+
+;; KSLRA 8|16|32
+(define_insn "kslra<VECI:mode><X:mode>"
+  [(set (match_operand:VECI 0 "register_operand"                  "=r")
+	(if_then_else:VECI
+	  (lt:X (match_operand:X 2 "register_operand"             " r")
```

Comment:
Incorrect check.
If my register contains 0x00000010, it is a negative value in terms of kslra16 but obviously a positive value in general meaning (i.e. gcc thinks that it `lt X, 0` if false).